docs/design/tanka.md

```diff
@@ -0,0 +1,223 @@
+challenging to model advanced requirements, for example deploying the same
+application to multiple clusters or to different environments.
+
+This usually leads to duplication on two level: The obvious one is the
+application level. Once an app needs to redeployed to another environment
+(`dev`, `prod`, etc.), it usually shares most of the configuration, except for
+some edge-cases (secrets, etc). `kubectl` does not really provide a utility to
+fully address this.
+
+Even more duplication happens on the systems level. Today, multiple applications
+are usually composed into a larger picture.  
+But deploying common building blocks as `postgresql` or `nginx` requires the
+same code (`Deployment`, `Service`, etc) that nearly everyone attempting to use
+these will write.
+
+While maintaining the same code for `dev` and `prod` might go well, it becomes
+unconquerable once it comes to multi-region (5+) deployments or even more
+versatile use-cases.
+
+#### Existing solutions
+Historically, this problem has been approached using string-templating (`helm`).
+While `helm` allows code-reuse using `values.yml`, a chart is only able to provide what
+has been thought of during writing. Even worse, charts are maintained inside of
+the `helm/charts` repository, which makes quick or even domain-specific edits
+hard. It is impossible to easily address edge-cases.
+
+A solution to this problem is called [jsonnet](https://jsonnet.org): It is
+basically `json` but with variables, conditionals, arithmetic, functions,
+**imports**, and error propagation and especially very clever [**deep-merging**](#edge-cases).
+
+#### Prior art
+Especially [ksonnet](https://ksonnet.io) had a big impact on this idea.
+While `ksonnet` really proved the idea to be working, is was based on the
+concept of components, building blocks consisting of prototypes and parameters,
+which may be composed into applications or modules which in turn may be applied
+to multiple environments.
+
+We believe such a concept overcomplicates the immediate goal of reducing
+duplication while allowing edge-cases, because it handles these on a higher
+conceptual level, instead of reusing the native capabilities of jsonnet.
+
+Code-reuse and composability is
+adequately provided by the native `import` feature of `jsonnet`. Sharing code
+beyond application boundaries is already enabled by `jsonnet-bundler`.
+
+While it is possible to mimic environments with native `jsonnet`, it falls short
+when it comes to actually applying it to the correct cluster. The raw `json`
+being returned by the compiler needs reconciling and it must be made sure it is
+applied to the correct cluster.
+
+This leaves us effectively with **`jsonnet`** and **[Environments](#environments)**
+
+## Code Reuse
+By using [`jsonnet`](https://jsonnet.org) as the underlying data templating
+language, tanka dynamic reusing of code, just like real programming languages do:
```

Comment:
"tanka dynamic reusing of code". huh? What does this mean?

Reply:
typo